### PR TITLE
Remove two iterators from Concat

### PR DESCRIPTION
**Pull request by jamesqo**
Was part of https://github.com/dotnet/corefx/pull/15389 but moved to another PR. Fixes https://github.com/dotnet/corefx/issues/15942

Changes:

- Remove 2 iterators from `Concat` that specialize for ICollections. Instead of having different iterators, set a flag on the iterator if all inputs so far have been ICollections. This allows us to determine in constant time whether we can get the count cheaply and preallocate for ToArray/ToList.

- This allows us to move the source enumerables into linked list nodes that are of uniform type, `ConcatNIterator<TSource>`. Previously, we had to typecast every time we crossed a node.

- Implement `GetCount` and `ToArray` in the derived classes rather than the base to avoid virtual calls and improve time complexity.
  - Avoid walking the linked list quadratically for `GetCount`, since we don't need to sum the counts in order.
  - Remove unnecessary virtual calls to `GetEnumerable` for 1 Concat followed by `ToArray`.
  - Make 1 Concat followed by `Count` cheap if the both of the concatees' counts can be gotten cheaply.

- Add a `ReserveOrAdd` method to `SparseArrayBuilder` which either reserves space for the enumerable if its count can be predetermined, or eagerly adds it otherwise. This saves quite a few lines of code.

**Performance tests:** <s>[memory leak reduction](https://gist.github.com/jamesqo/c8ccc43c2e9a7b8bb35c6cc6bebeb9ec) / [regressions from this change](https://gist.github.com/jamesqo/bde594c796e2c07117f77e88239a2c40)</s> Need to be updated.

/cc @JonHanna @stephentoub @VSadov

### CODE REVIEW

**Comment by karelz** [comment]
cc: @ianhays @safern as it touches also Collections.

**Comment by jamesqo** [comment]
@karelz Not really. Those are just Common.files which are only used by Linq.

**Comment by karelz** [comment]
Ah, thanks for clarification. Will try to remember it in future :)

**Comment by jamesqo** [comment]
Test Innerloop Ubuntu14.04 Release Build and Test

File: src/System.Linq/src/System/Linq/SelectMany.cs [hunk 1]
```diff
@@ -231,18 +231,11 @@ public TResult[] ToArray()
                 {
                     IEnumerable<TResult> enumerable = _selector(element);
 
-                    int count;
-                    if (EnumerableHelpers.TryGetCount(enumerable, out count))
+                    // TODO: Add test to ensure the bug is fixed.
+                    if (builder.ReserveOrAdd(enumerable))
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/SelectMany.cs]: Please submit a separate PR with the SelectMany bug fix.  It doesn't have anything to do with Concat.

File: src/System.Linq/src/System/Linq/Concat.cs [hunk 2]
```diff
@@ -226,182 +101,188 @@ internal override IEnumerable<TSource> GetEnumerable(int index)
...
+        ///
+        /// This linked-list implementation means that rather than each MoveNext/Current call having to
+        /// traverse all of the previous sources, we only have to traverse all of the previous sources
+        /// once per chained enumerable.  An alternative would be to use an array to store all of the
+        /// enumerables, but this has a much better memory profile and without much additional run-time cost.
+        ///
+        /// At face value, using list nodes seems worse than chaining together the iterators directly and
+        /// letting the iterator itself hold a reference to the latest source. So why allocate a list node in
+        /// addition to each iterator? It lets the GC reclaim intermediary iterators in the chain which
+        /// have fields that are not relevant to later iterators, like <see cref="_enumerator"/> or
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Concat.cs]: Is this really a win in _real_ usage?  We're allocating additional objects in order to allow the GC to reclaim other objects.  And in general, the objects you build up in these chains aren't held onto for very long: you do a few concats and enumerate the result.

**Comment by stephentoub** [comment]
> Performance tests: memory leak reduction / regressions from this change

I'm not seeing this as an improvement.  The test for letting things get GC'd is creating a chain of 5M concats... that's not a real scenario.  The cases shown in the regressions list are more representative of real usage and are regressions.

File: src/System.Linq/src/System/Linq/SelectMany.cs (hunk 1 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/SelectMany.cs]: Sure

File: src/System.Linq/src/System/Linq/Concat.cs (hunk 2 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Concat.cs]: @stephentoub There are other aspects where this is a win. We don't pay for typecasting to ICollection up front in Concat anymore, it is deferred until ToArray/ToList is called and iteration tends to be more common. `GetEnumerable` also becomes a [lot faster and simpler](https://github.com/dotnet/corefx/pull/15943/files#diff-d8b3be41edbd1e2e3c65380665467a2dR207) because the nodes are uniform type, so we don't have to typecast with O(N^2) behavior.

**Comment by jamesqo** [comment]
@stephentoub I'm still in the process of looking thru the regressions, but one of the reasons that things like `Count` / `ToList` seem a lot slower especially for small counts is because [we call into `TryGetCount`](https://github.com/dotnet/corefx/pull/15943/files#diff-d8b3be41edbd1e2e3c65380665467a2dR67) which tries to typecast to both ICollection and IIListProvider before giving up. This is a tradeoff that is well worth it because we no longer say we can't get the count cheaply if one of our inputs is an IIListProvider. [This benchmark runs a lot faster.](https://gist.github.com/jamesqo/14c961d47838e6d6b886f0f07ce1aa89)

Another reason why the previous implementation seems so much faster is because it does its typecasting when `Concat` is called, while the new implementation defers it until `GetCount`. So really there's not much of a difference besides the extra typecast.

**Comment by jamesqo** [comment]
@stephentoub The regressions for `ToArray` are for the same reason; we're trying to typecast to IIListProvider too, instead of just ICollection. So the new implementation is going to reduce GC allocations for IIListProviders because we're going to preallocate based on their count.

It should also be noted that most of the regressions are exaggerated because the typecast is being moved from when `Concat` is called to GetCount/ToArray/ToList. The benchmark only calls those methods in a loop, not Concat. So it's including the cost of the typecasts in the new implementation, but not the old one.

**Comment by jamesqo** [comment]
@stephentoub, I've worked on this some more and you might want to take another look. **I've removed the `SingleLinkedNode` thing for the time being, so we only make 1 allocation per Concat.** I've posted some more comprehensive benchmarks [here](https://gist.github.com/jamesqo/c521977cf2423b4eae9d6109a5d718e6) that demonstrate both the improvements/regressions introduced by this PR. Some things to note:

- Before, we were saying that `x.Concat(y)` could not be counted cheaply even if x and y were both IIListProviders. This meant we did not preallocate for `ToList` and there was wasteful memory usage. This PR fixes that for the case of 1 Concat [and decreases GCs a lot](https://gist.github.com/jamesqo/c521977cf2423b4eae9d6109a5d718e6#file-analysis-L1063-L1092).

- The results at the top of the analysis show pretty drastic regressions if all inputs are ICollections. However, these results are actually skewed in favor of the old implementation; the old implementation typecasts to `ICollection` before the Concat iterator is returned, so that doesn't get included in the benchmark. The new implementation doesn't do the typecast until GetCount/ToArray/ToList is called, so it does get included. So many of the regressions are exaggerated.

File: src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs [hunk 3]
```diff
@@ -176,6 +176,33 @@ public void Reserve(int count)
...
         /// <summary>
+        /// Reserves a region if the items' count can be predetermined; otherwise, adds the items to this builder.
+        /// </summary>
+        /// <param name="items">The items to reserve or add.</param>
+        /// <returns><c>true</c> if the items were reserved; otherwise, <c>false</c>.</returns>
+        /// <remarks>
+        /// If the items' count is predetermined to be 0, no reservation is made and the return value is <c>false</c>.
+        /// The effect is the same as if the items were added, since adding an empty collection does nothing.
+        /// </remarks>
+        public bool ReserveOrAdd(IEnumerable<T> items)
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs]: Unit tests for this?

File: src/System.Linq/src/System/Linq/SelectMany.cs [hunk 4]
```diff
@@ -231,18 +231,10 @@ public TResult[] ToArray()
                 {
                     IEnumerable<TResult> enumerable = _selector(element);
 
-                    int count;
-                    if (EnumerableHelpers.TryGetCount(enumerable, out count))
+                    if (builder.ReserveOrAdd(enumerable))
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/SelectMany.cs]: Why is this change to SelectMany in this PR for Concat?

File: src/System.Linq/src/System/Linq/Concat.cs [hunk 5]
```diff
@@ -21,201 +21,83 @@ public static IEnumerable<TSource> Concat<TSource>(this IEnumerable<TSource> fir
...
-                return new Concat2EnumerableIterator<TSource>(_first, _second);
-            }
+            public override Iterator<TSource> Clone() => new Concat2Iterator<TSource>(_first, _second);
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
-                return new ConcatNEnumerableIterator<TSource>(this, next, 2);
+                // Instead of linking directly to this Concat2Iterator, we create 2 new ConcatNIterators
+                // for the first two sources and create a third one to hold the next source.
+                // This simplifies `GetEnumerable` because the nodes are of uniform type and we don't
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Concat.cs]: > This simplifies 'GetEnumerable'

And makes the common case of `foreach (var item in source.Concat(something).Concat(somethingElse))` more expensive, no?

**Comment by stephentoub** [comment]
@jamesqo, thanks for your work on this, but I'm still not understanding why this PR is actually a good thing.  The most common scenario for Concat is to have a few concats that are then enumerated, and it seems this PR makes that slower, no?

**Comment by jamesqo** [comment]
> thanks for your work on this, but I'm still not understanding why this PR is actually a good thing. The most common scenario for Concat is to have a few concats that are then enumerated, and it seems this PR makes that slower, no?

The main point of this PR was to remove the two iterators, but I guess I got a little sidetracked with the uniform-type node thing. I thought the previous code for looping over the enumerables looked pretty verbose with all of the typecasts, so that's why I wanted to change that. Nevertheless, I've reverted back to the original scheme where the tail is an abstract class, and fortunately I found a way to loop without too much boilerplate by enscapulating the typecast in a property.

File: src/System.Linq/src/System/Linq/Concat.cs (hunk 5 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/Concat.cs]: Fixed

File: src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs]: I will add them when I get the chance.

File: src/System.Linq/src/System/Linq/SelectMany.cs (hunk 4 above)
Comment (jamesqo) [inline src/System.Linq/src/System/Linq/SelectMany.cs]: I introduced this method to help with `Concat` (it's used there) then I realized it could be used in `SelectMany` too.

File: src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/SparseArrayBuilder.cs]: @stephentoub Nvm, this type actually depends on `EnumerableHelpers.Linq.cs` which makes use of internal Linq interfaces. So there's no way to write unit tests for this type currently, unless maybe with `InternalsVisibleTo`. But either way that should probably be done in a separate PR bc some existing methods are outside the scope of this PR.

**Comment by jamesqo** [comment]
Updated the gist again with latest commit.

**Comment by danmosemsft** [comment]
@VSadov @OmarTawfik @stephentoub seems this is ready for next review.

**Comment by karelz** [comment]
@VSadov @OmarTawfik ping again
